src/tunnel/tcp.rs

```diff
@@ -72,45 +72,39 @@ pub async fn tcp_proxy_server(
 }
 
 /// Handles a new TCP connection with its assigned virtual port.
-async fn handle_tcp_proxy_connection(
-    mut socket: TcpStream,
+pub async fn handle_tcp_proxy_connection<A>(
+    mut socket: A,
     virtual_port: VirtualPort,
     port_forward: PortForwardConfig,
     bus: Bus,
-) -> anyhow::Result<()> {
+) -> anyhow::Result<()>
+where
+    A: AsyncReadExt + AsyncWriteExt + Unpin
+{
     let mut endpoint = bus.new_endpoint();
     endpoint.send(Event::ClientConnectionInitiated(port_forward, virtual_port));
 
-    let mut buffer = Vec::with_capacity(MAX_PACKET);
+    let mut buffer = vec![0; MAX_PACKET];
+
     loop {
         tokio::select! {
-            readable_result = socket.readable() => {
```

Comment:
I'm worried some of these changes will affect the performance of the default `TCP` port-forward; have you done extensive testing with the rest of onetun's features?

I understand this change was made to make it compile with the `AsyncReadExt + AsyncWriteExt + Unpin` traits. It would also be possible to make the HTTP proxy use its own tunnel code entirely (while sharing the same port pool).